simplicity-sys/src/bitstream.rs

```diff
@@ -0,0 +1,49 @@
+// Simplicity Bindings
+// Written in 2022 by
+//   Christian Lewe <clewe@blockstream.com>
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+use libc::{c_char, c_int, c_uchar, FILE};
+use std::marker::PhantomData;
+
+/// Stream of bits.
+///
+/// Corresponds to `bitstream` in `depend/simplicity/bitstream.h`.
+#[repr(C)]
+pub struct Bitstream<'a> {
+    file: *mut FILE,
+    available: c_int,
+    byte: c_uchar,
+    _marker: PhantomData<&'a u8>,
+}
+
+impl<'a> From<&'a [u8]> for Bitstream<'a> {
+    fn from(bytes: &'a [u8]) -> Self {
+        unsafe {
+            let mode = "r";
+            // `file` does not need to be freed as
+            // `bytes` will free itself at the end of its lifetime
```

Comment:
In 213c69795b5c00d46bd593c3d7af764fa539e973:

This comment doesn't make sense to me. At some point you do need to `fclose` `file`, don't you?

Reply:
`Bitstream<'a>` lives at most as long as `&'a bytes`. According to man pages, `fmemopen` does not open a file descriptor, nor does a read-only stream have any buffers that need to be flushed. `fclose` simply closes file descriptors and flushes. If I understand correctly, this means that the `FILE` pointer in the stream can safely be dropped without freeing it.

That being said, not freeing a `FILE` pointer is probably bad practice. I will add a destructor.

Reply:
I'm sure that you need to actually call `fclose` at some point in this source file.

Otherwise, the lifetime stuff looks like it's been fixed.

Reply:
Actually, on further reading I don't know. I can't find a citation either way. It is surprising that there is no descriptor and that `fileno` will just fail.